crates/builder/src/graph/adj_list.rs

```diff
@@ -32,25 +39,51 @@ impl<NI: Idx, EV> AdjacencyList<NI, EV> {
         NI: Send + Sync,
         EV: Send + Sync,
     {
-        NI::new(self.edges.par_iter().map(|v| v.len()).sum())
+        NI::new(
+            self.edges
+                .par_iter()
+                .map(|v| v.lock().expect("Could not acquire lock").len())
+                .sum(),
+        )
     }
 
     #[inline]
     pub(crate) fn degree(&self, node: NI) -> NI {
-        NI::new(self.edges[node.index()].len())
+        NI::new(
+            self.edges[node.index()]
+                .lock()
+                .expect("Could not aqcuire lock")
+                .len(),
+        )
     }
-}
 
-impl<NI: Idx, EV> AdjacencyList<NI, EV> {
     #[inline]
-    pub(crate) fn targets_with_values(&self, node: NI) -> &[Target<NI, EV>] {
-        self.edges[node.index()].as_slice()
+    pub(crate) fn insert(&self, source: NI, target: Target<NI, EV>) {
+        let edges = &mut self.edges[source.index()]
```

Comment:
You don't really need the `&mut` here, no?

```suggestion
        let mut edges = self.edges[source.index()]
```